### PR TITLE
Landing page builder feature

### PR DESCRIPTION
**Pull request by damcou**

### CODE REVIEW

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+            <field name="date_from">
+                <argument name="data" xsi:type="array">
+                    <item name="config" xsi:type="array">
+                        <item name="additionalClasses" xsi:type="string">admin__field-date</item>
+                        <item name="source" xsi:type="string">landing_page</item>
+                        <item name="formElement" xsi:type="string">date</item>
+                        <item name="dataType" xsi:type="string">text</item>
+                        <item name="label" xsi:type="string" translate="true">Timeframe From</item>
+                        <item name="dataScope" xsi:type="string">date_from</item>
+                        <item name="notice" xsi:type="string" translate="true">If you don't set any timeframe, the landing page will be accessible indefinitely.</item>
```
Comment (AlexEven) [inline view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: Define a timeframe when the landing page should be published (for a specific campaign for example). If no timeframe is defined, the landing page will be online unless you unpublish it manually.

* When the timeframe is over, do we unpublish the page? How does it work?

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+                </argument>
+            </field>
+        </container>
+        <field name="custom_js">
+            <argument name="data" xsi:type="array">
+                <item name="config" xsi:type="array">
+                    <item name="source" xsi:type="string">landing_page</item>
+                    <item name="dataType" xsi:type="string">text</item>
+                    <item name="formElement" xsi:type="string">textarea</item>
+                    <item name="label" xsi:type="string" translate="true">Javascript</item>
```
Comment (AlexEven) [inline view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: Custom Javascript

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+                    <item name="tooltip" xsi:type="array">
+                        <item name="description" xsi:type="string">You can add some Javascript code on the top of the page according to your needs.</item>
```
Comment (AlexEven) [inline view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: We can remove the tooltip, it's redundant with the placeholder

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+                    </item>
+                    <item name="placeholder" xsi:type="string" translate="true">Add javascript to enchance your page even more. Paste your code here.</item>
```
Comment (AlexEven) [inline view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: Add some Javascript code to enhance your page.

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+            </argument>
+        </field>
+        <field name="custom_css">
+            <argument name="data" xsi:type="array">
+                <item name="config" xsi:type="array">
+                    <item name="source" xsi:type="string">landing_page</item>
+                    <item name="dataType" xsi:type="string">text</item>
+                    <item name="formElement" xsi:type="string">textarea</item>
+                    <item name="label" xsi:type="string" translate="true">Custom CSS</item>
+                    <item name="placeholder" xsi:type="string" translate="true">Add Custom CSS</item>
```
Comment (AlexEven) [inline view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: Add custom CSS to further improve the style of your page.

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+                    <item name="tooltip" xsi:type="array">
+                        <item name="description" xsi:type="string">You can add some custom CSS on the top of the page according to your needs.</item>
```
Comment (AlexEven) [inline view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: We can remove this tooltip too.

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+                    <item name="dataScope" xsi:type="string">custom_css</item>
+                </item>
+            </argument>
+        </field>
+    </fieldset>
+    <fieldset name="content">
+        <argument name="data" xsi:type="array">
+            <item name="config" xsi:type="array">
+                <item name="collapsible" xsi:type="boolean">true</item>
+                <item name="label" xsi:type="string" translate="true">CMS Content</item>
```
Comment (AlexEven) [inline view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: CMS editor

Comment (AlexEven) [reply view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: Can we add a tool tip here? There's a blue box in Lucia's design

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+                    <item name="additionalClasses" xsi:type="string">admin__field-wide</item>
+                </item>
+            </argument>
+        </field>
+    </fieldset>
+    <fieldset name="search_configuration">
+        <argument name="data" xsi:type="array">
+            <item name="config" xsi:type="array">
+                <item name="collapsible" xsi:type="boolean">true</item>
+                <item name="label" xsi:type="string" translate="true">Search Configuration (query, filters, product ordering)</item>
```
Comment (AlexEven) [inline view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: Merchandises products

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+            </item>
+        </argument>
+        <field name="query">
+            <argument name="data" xsi:type="array">
+                <item name="config" xsi:type="array">
+                    <item name="source" xsi:type="string">landing_page</item>
+                    <item name="dataType" xsi:type="string">text</item>
+                    <item name="formElement" xsi:type="string">input</item>
+                    <item name="label" xsi:type="string" translate="true">Query</item>
```
Comment (AlexEven) [inline view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: Search query

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+                </item>
+            </argument>
+        </field>
+        <field name="configuration">
+            <argument name="data" xsi:type="array">
+                <item name="config" xsi:type="array">
+                    <item name="source" xsi:type="string">landing_page</item>
+                    <item name="dataType" xsi:type="string">text</item>
+                    <item name="formElement" xsi:type="string">input</item>
+                    <item name="label" xsi:type="string" translate="true">Configuration</item>
```
Comment (AlexEven) [inline view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: Where is this displayed?

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml (hunk 9 above)
Comment (AlexEven) [reply view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: Add placeholder: "Enter the focus keyword of your landing page"

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+            <argument name="data" xsi:type="array">
+                <item name="config" xsi:type="array">
+                    <item name="source" xsi:type="string">landing_page</item>
+                    <item name="label" xsi:type="string" translate="true">Unique URL key</item>
+                    <item name="dataType" xsi:type="string">text</item>
+                    <item name="formElement" xsi:type="string">input</item>
+                    <item name="dataScope" xsi:type="string">url_key</item>
+                    <item name="notice" xsi:type="string" translate="true">Please make sure that the chosen URL isn't already used by the Magento's URL rewrite system (including products, categories and CMS pages)</item>
+                    <item name="tooltip" xsi:type="array">
+                        <item name="description" xsi:type="string">The URL key determine which URL the landing will have (ex : http://www.mystore.com/url-to-the-landing-page)</item>
```
[the remark this answers is not in the document]
Comment (damcou) [reply view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: 1) Unfortunately no, the tooltip can only be handled in the XML
2) Yes, you can use whatever link structure you want ;) 
3) Would be difficult, and also not really meaningful cause the url depends on the store, and by default the landing page is applied on all stores (so different urls are generated)

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+            </argument>
+        </field>
+        <field name="meta_description">
+            <argument name="data" xsi:type="array">
+                <item name="config" xsi:type="array">
+                    <item name="source" xsi:type="string">landing_page</item>
+                    <item name="dataType" xsi:type="string">text</item>
+                    <item name="formElement" xsi:type="string">textarea</item>
+                    <item name="tooltip" xsi:type="array">
+                        <item name="description" xsi:type="string">The meta description can be customized to enhance your SEO.</item>
```
[the remark this answers is not in the document]
Comment (damcou) [reply view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: Meta description is not set

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml (hunk 1 above)
Comment (damcou) [reply view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: Before displaying the page, Magento checks the timeframe, so even if the page is not marked as "unpublished", it won't be displayed if it's out of the timeframe.

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml (hunk 10 above)
Comment (damcou) [reply view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: It's not ;)

File: view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml [hunk 13]
```diff
@@ -0,0 +1,305 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<form xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Ui:etc/ui_configuration.xsd">
+    <argument name="data" xsi:type="array">
+        <item name="js_config" xsi:type="array">
+            <item name="provider" xsi:type="string">algolia_algoliasearch_landingpage_form.landingpage_form_data_source</item>
+            <item name="deps" xsi:type="string">algolia_algoliasearch_landingpage_form.landingpage_form_data_source</item>
+        </item>
+        <item name="label" xsi:type="string" translate="true">Main Information</item>
```
[the remark this answers is not in the document]
Comment (damcou) [reply view/adminhtml/ui_component/algolia_algoliasearch_landingpage_form.xml]: It's not displayed

File: Controller/Adminhtml/Landingpage/Delete.php [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+        /** @var \Magento\Backend\Model\View\Result\Redirect $resultRedirect */
+        $resultRedirect = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
+
+        $landingPageId = $this->getRequest()->getParam('id');
+        if ($landingPageId) {
+            try {
+                /** @var \Algolia\AlgoliaSearch\Model\LandingPage $landingPage */
+                $landingPage = $this->landingPageFactory->create();
+                $landingPage->getResource()->load($landingPage, $landingPageId);
+                $landingPage->getResource()->delete($landingPage);
```
Comment (JanPetr) [inline Controller/Adminhtml/Landingpage/Delete.php]: URL is deleted automatically?

File: Controller/Adminhtml/Landingpage/Index.php [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        $breadMain = __('Algolia | Landing Page Builder');
+
+        /** @var \Magento\Framework\View\Result\Page $resultPage */
+        $resultPage = $this->resultFactory->create(ResultFactory::TYPE_PAGE);
+        $resultPage->setActiveMenu('Algolia_AlgoliaSearch::manage');
+        $resultPage->getConfig()->getTitle()->prepend($breadMain);
+
+        $dataPersistor = $this->_objectManager->get(\Magento\Framework\App\Request\DataPersistorInterface::class);
+        $dataPersistor->clear('landing_page');
```
Comment (JanPetr) [inline Controller/Adminhtml/Landingpage/Index.php]: What is data persistor?

File: Controller/Adminhtml/Landingpage/Delete.php (hunk 14 above)
Comment (damcou) [reply Controller/Adminhtml/Landingpage/Delete.php]: Yes, I added a _afterDelete() method on the resource model : https://github.com/algolia/algoliasearch-magento-2/blob/feat-landing-page-builder/Model/ResourceModel/LandingPage.php#L79

File: Controller/Adminhtml/Landingpage/Index.php (hunk 15 above)
Comment (damcou) [reply Controller/Adminhtml/Landingpage/Index.php]: I guess it's used to keep the submitted data when needed : https://github.com/algolia/algoliasearch-magento-2/blob/feat-landing-page-builder/Controller/Adminhtml/Landingpage/Save.php#L119.
I took inspiration from the M2 CMS page feature for this : https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Cms/Controller/Adminhtml/Page/Index.php#L56

**Pull request merged**